tree/ntuple/v7/inc/ROOT/RClusterPool.hxx

```diff
@@ -0,0 +1,128 @@
+   struct RWorkItem {
+      std::promise<std::unique_ptr<RCluster>> fPromise;
+      DescriptorId_t fClusterId = kInvalidDescriptorId;
+      RPageSource::ColumnSet_t fColumns;
+      RWorkItem() = default;
+   };
+
+   /// Clusters that are currently being processed by the I/O thread.  Every in flight cluster has a corresponding
+   /// work item.
+   struct RInFlightCluster {
+      std::future<std::unique_ptr<RCluster>> fFuture;
+      DescriptorId_t fClusterId = kInvalidDescriptorId;
+      RPageSource::ColumnSet_t fColumns;
+      /// By the time a cluster has been loaded, it might not be necessary anymore.
+      bool fIsExpired = false;
+      RInFlightCluster() = default;
+      bool operator== (const RInFlightCluster &other) const { return fClusterId == other.fClusterId && fColumns == other.fColumns; }
+      bool operator!= (const RInFlightCluster &other) const { return !(*this == other); }
+      /// First order by cluster id, then by number of columns, than by the column ids in fColumns
+      bool operator< (const RInFlightCluster &other) const;
+   };
+
+   RPageSource *fPageSource;
+   /// The number of clusters before the currently active cluster that should stay in the pool if present
+   unsigned int fWindowPre;
+   /// The number of desired clusters in the pool, including the currently active cluster
+   unsigned int fWindowPost;
+   /// The cache of clusters around the currently active cluster
+   std::vector<std::shared_ptr<RCluster>> fPool;
+
+   std::mutex fLockInFlightClusters;
+   /// The clusters that were handed off to the I/O thread
+   std::vector<RInFlightCluster> fInFlightClusters;
+
+   std::mutex fLockWorkQueue;
+   std::condition_variable fCvHasWork;
+   /// The communication channel to the I/O thread
+   std::queue<RWorkItem> fWorkQueue;
+
+   std::thread fThreadIo;
+
+   /// Every cluster id has at most one corresponding RCluster pointer in the pool
+   std::shared_ptr<RCluster> FindInPool(DescriptorId_t clusterId);
+   /// Returns an index of an unused element in fPool; caller has to ensure that a free slot exists
+   size_t FindFreeSlot();
+   /// The I/O thread routine, there is exactly one I/O thread in flight for every cluster pool
+   void ExecLoadClusters();
+   /// Returns the given cluster from the pool, which needs to contain at least the columns `columns`.
+   /// Executed at the end of GetCluster when all missing data pieces have been sent to the load queue.
+   /// Ideally, the function returns without blocking if the cluster is already in the pool.
+   std::shared_ptr<RCluster> WaitFor(DescriptorId_t clusterId, const RPageSource::ColumnSet_t &columns);
+
+public:
+   static const unsigned int kDefaultPoolSize = 4;
+   RClusterPool(RPageSource *pageSource, unsigned int size);
+   explicit RClusterPool(RPageSource *pageSource) : RClusterPool(pageSource, kDefaultPoolSize) {}
+   explicit RClusterPool(const RClusterPool &other) = delete;
```

Comment:
Is the use of `explicit` intentional? It might make sense, I've just never seen it.